### PR TITLE
IBX-5705: OAuth server

### PR DESCRIPTION
**Pull request by adriendupuis**
| Question      | Answer
| ------------- | ---
| JIRA Ticket   | [IBX-5705](https://issues.ibexa.co/browse/IBX-5705)
| Versions      | 4.6
| Edition       | All

- Document the OAuth 2 server
- Merge and revamp OAuth client doc

#### Checklist

- [ ] Text renders correctly
- [ ] Text has been checked with vale
- [ ] Description metadata is up to date
- [ ] Redirects cover removed/moved pages
- [ ] Code samples are working
- [ ] PHP code samples have been fixed with PHP CS fixer
- [ ] Added link to this PR in relevant JIRA ticket or code PR

### CODE REVIEW

**Review by Nattfarinn — APPROVED** [review]

File: docs/users/oauth_server.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        #            - Ibexa\OAuth2Server\Security\Guard\OAuth2Authenticator
+        oauth2_token:
+            pattern: ^/token$
+            security: false
+```
+
+## Resource Server configuration
+
+To allow resource routes to be accessible through OAuth authorization,
+you must define a firewall using the `Ibexa\OAuth2Server\Security\Guard\OAuth2Authenticator`.
```
Comment (dabrt) [inline docs/users/oauth_server.md:107]:
```suggestion
you must define a firewall by using `Ibexa\OAuth2Server\Security\Guard\OAuth2Authenticator`.
```

File: docs/users/oauth_server.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+The following firewall example allows the REST API to be accessed as an OAuth resource.
+It must be before firewall with less restrictuve pattern like `ibexa_front`.
```
Comment (dabrt) [inline docs/users/oauth_server.md:110]:
```suggestion
It must be placed before the firewall with a less restrictive pattern like `ibexa_front`.
```

File: docs/users/oauth_server.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+                    - Ibexa\OAuth2Server\Security\Guard\OAuth2Authenticator
+                entry_point: Ibexa\OAuth2Server\Security\Guard\OAuth2Authenticator
+            stateless: true
+
+        ibexa_front:
+            pattern: ^/
+            #…
+```
+
+TODO: Can it break same-domain Back Office like with [HTTP basic authentication](rest_api_authentication.md#configuration)?
```
Comment (dabrt) [inline docs/users/oauth_server.md:132]: Add information here

File: docs/users/oauth_server.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+## Client
+
+### Add a client
+
+You need the client redirect URIs to create it.
```
Comment (dabrt) [inline docs/users/oauth_server.md:138]:
```suggestion
You need the client redirect URIs to create a client.
```

File: docs/users/oauth_server.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+Use `league:oauth2-server:create-client` command to create a client.
+For example:
+
+```bash
+php bin/console league:oauth2-server:create-client 'Example OAuth2 Client' example-oauth2-client 9876543210987654321098765432109876543210 --scope=default \
+  --redirect-uri=https://example.com/oauth2-callback
+```
+
+`--redirect-uri` can be used multiple time.
```
Comment (dabrt) [inline docs/users/oauth_server.md:150]:
```suggestion
!!! note
   You can call `--redirect-uri` multiple times.
```

File: docs/users/oauth_server.md [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+Redirect URIs can be added after the creation with `league:oauth2-server:update-client` command.
```
Comment (dabrt) [inline docs/users/oauth_server.md:152]:
```suggestion
Alternatively, you could add redirect URIs after you create a client, by using the `league:oauth2-server:update-client` command.
```

File: docs/users/oauth_server.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+For example:
+
+```bash
+php bin/console league:oauth2-server:update-client example-oauth2-client \
+  --add-redirect-uri=https://example.com/another-oauth2-callback
+```
+
+There is also commands to list all the configured clients (`league:oauth2-server:list-clients`),
+or to delete a client (`league:oauth2-server:delete-client`).
```
Comment (dabrt) [inline docs/users/oauth_server.md:161]:
```suggestion
Other commands let you list all the configured clients (`league:oauth2-server:list-clients`)
or delete a client (`league:oauth2-server:delete-client`).
```

File: docs/users/oauth_server.md [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+
+- In a terminal, see `bin/console list league:oauth2-server` for a list of all the commands to maintain your clients.
+  Use `bin/console help <command>` to usage detail for each of them.
```
Comment (dabrt) [inline docs/users/oauth_server.md:164]:
```suggestion
!!! note

    For a list of all the commands that you can use maintain your clients, in a terminal, run `bin/console list league:oauth2-server`.
    To see usage details for each of the commands, run `bin/console help <command>` . 
```

File: docs/users/oauth_server.md [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+or to delete a client (`league:oauth2-server:delete-client`).
+
+- In a terminal, see `bin/console list league:oauth2-server` for a list of all the commands to maintain your clients.
+  Use `bin/console help <command>` to usage detail for each of them. 
+- See [package's online documentation](https://github.com/thephpleague/oauth2-server-bundle/blob/master/docs/basic-setup.md).
+
+!!! note
+
+    The `ìdentifier` and the `secret` will be needed by the client settings.
+
```
Comment (dabrt) [inline docs/users/oauth_server.md:170]:
```suggestion
```
This information is duplicated in the next section

File: docs/users/oauth_server.md [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+php bin/console league:oauth2-server:update-client example-oauth2-client \
+  --add-redirect-uri=https://example.com/another-oauth2-callback
+```
+
+There is also commands to list all the configured clients (`league:oauth2-server:list-clients`),
+or to delete a client (`league:oauth2-server:delete-client`).
+
+- In a terminal, see `bin/console list league:oauth2-server` for a list of all the commands to maintain your clients.
+  Use `bin/console help <command>` to usage detail for each of them. 
+- See [package's online documentation](https://github.com/thephpleague/oauth2-server-bundle/blob/master/docs/basic-setup.md).
```
Comment (dabrt) [inline docs/users/oauth_server.md:165]:
```suggestion

    For more information, see the package's [online documentation](https://github.com/thephpleague/oauth2-server-bundle/blob/master/docs/basic-setup.md).
```

**Review by dabrt — CHANGES_REQUESTED** [review]

File: code_samples/user_management/oauth_google/config/packages/knpu_oauth2_client.yaml [hunk 11]
```diff
@@ -1,11 +1,10 @@
 knpu_oauth2_client:
     clients:
-        # Configure your clients as described here: https://github.com/knpuniversity/oauth2-client-bundle#configuration
+        # configure your clients as described here: https://github.com/knpuniversity/oauth2-client-bundle#configuration
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply code_samples/user_management/oauth_google/config/packages/knpu_oauth2_client.yaml:3]: This is capitalized this way in the original file.

File: docs/users/oauth_client.md [hunk 12]
```diff
@@ -0,0 +1,135 @@
+---
+description: Allow users to log into Ibexa DXP through external OAuth2 authorization servers.
+---
+
+# OAuth client
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply docs/users/oauth_client.md:5]: Done in e9d96fffd3f35921d0f6a446e4b610ce489cad4b

File: docs/users/oauth_server.md [hunk 13]
```diff
@@ -0,0 +1,178 @@
+---
+description: Other applications can authenticate Ibexa DXP users through OAuth2 protocol then access to their resources on the platform.
+---
+
+# OAuth Server
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply docs/users/oauth_server.md:5]: Done in e9d96fffd3f35921d0f6a446e4b610ce489cad4b

File: docs/users/oauth_server.md [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+
+    ```bash
+    mysql -u <username> -p <password> <database_name> -e 'SET FOREIGN_KEY_CHECKS=0;'
+    php bin/console ibexa:doctrine:schema:dump-sql vendor/ibexa/oauth2-server/src/bundle/Resources/config/schema.yaml | mysql -u <username> -p <password> <database_name>
+    mysql -u <username> -p <password> <database_name> -e 'SET FOREIGN_KEY_CHECKS=1;'
+    ```
+
+=== "PostgreSQL"
+
+    php bin/console ibexa:doctrine:schema:dump-sql --force-platform=postgres vendor/ibexa/oauth2-server/src/bundle/Resources/config/schema.yaml | psql <database_name>
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply docs/users/oauth_server.md:46]: Fixed in 3f1b6595b860c630d4d1ae1b4333870b1efc67ad

File: docs/users/oauth_server.md [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+                    - Ibexa\OAuth2Server\Security\Guard\OAuth2Authenticator
+                entry_point: Ibexa\OAuth2Server\Security\Guard\OAuth2Authenticator
+            stateless: true
+
+        ibexa_front:
+            pattern: ^/
+            #…
+```
+
+TODO: Can it break same-domain Back Office like with [HTTP basic authentication](rest_api_authentication.md#configuration)?
```
Comment (adriendupuis) [inline docs/users/oauth_server.md:134]:
```suggestion
```

File: docs/users/oauth_client.md [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+composer require league/oauth2-google
+```
+
+### Enable OAuth2 client
+
+The client must be associated to a [SiteAccess scope](multisite_configuration.md#scope).
+
+In the following example, the previous OAuth2 client `google` is enabled for the `admin` SiteAccess:
+
+``` yaml
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply docs/users/oauth_client.md:44]: Effectively, the firewall example is a tall one.

[the remark this answers is not in the document]
Comment (adriendupuis) [reply docs/users/oauth_client.md:44]: @dabrt  I didn't manage to make it works

File: docs/users/oauth_client.md [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+    - `ResourceOwnerIdToUserMapper` doesn't create a new user, but loads a user (resource owner) based on their identifier.
+    - `ResourceOwnerEmailToUserMapper` doesn't create a new user, but loads a user (resource owner) based on their email.
+- `ResourceOwnerToExistingOrNewUserMapper` checks if the user exists and loads them if they do.
+  If they don't, the mapper creates a new user in the Repository.
+
+To use `ResourceOwnerToExistingOrNewUserMapper`, you need to extend it in your custom mapper.
+
+!!! tip "OAuth User content type"
+
+    When you implement your own mapper for external login,
```
Comment (adriendupuis) [inline docs/users/oauth_client.md:77]:
```suggestion
    When you implement your own mapper for external login,
```

File: docs/users/oauth_client.md [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+
+## Configure OAuth2 client
+
+### Configure connection to Authorization Server
+
+Details of the configuration depend on the OAuth2 Authorization Server that you want to use.
+For sample configurations for different providers,
+see [`knpuniversity/oauth2-client-bundle` configuration](https://github.com/knpuniversity/oauth2-client-bundle#configuration).
+Some client types require additional packages.
+(Missing package is, at least, indicated in an error message.)
```
Comment (julitafalcondusza) [inline docs/users/oauth_client.md:22]:
```suggestion
Missing package isnindicated in an error message.
```

File: docs/users/oauth_client.md [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+
+For example, the following configuration creates a `google` client for Google OAuth2 Authorization Server to log users in.
+The two environment variables `OAUTH_GOOGLE_CLIENT_ID` and `OAUTH_GOOGLE_CLIENT_SECRET`
```
Comment (julitafalcondusza) [inline docs/users/oauth_client.md:25]:
```suggestion
Two environment variables, `OAUTH_GOOGLE_CLIENT_ID` and `OAUTH_GOOGLE_CLIENT_SECRET`,
```

File: docs/users/oauth_client.md [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+correspond to [the set-up on Google side](https://support.google.com/cloud/answer/6158849).
+
+``` yaml
+[[= include_file('code_samples/user_management/oauth_google/config/packages/knpu_oauth2_client.yaml') =]]
+```
+
+To use the `google` client type, you must install the following package:
```
Comment (julitafalcondusza) [inline docs/users/oauth_client.md:32]:
```suggestion
To use the `google` client type, you need to install the following package:
```

File: docs/users/oauth_client.md [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+
+```bash
+composer require league/oauth2-google
+```
+
+### Enable OAuth2 client
+
+The client must be associated to a [SiteAccess scope](multisite_configuration.md#scope).
```
Comment (julitafalcondusza) [inline docs/users/oauth_client.md:40]:
```suggestion
The client needs to be a part of the [SiteAccess scope](multisite_configuration.md#scope).
```

File: docs/users/oauth_client.md [hunk 22]
```diff
@@ -0,0 +1,132 @@
...
+
+## Resource owner mappers
+
+Resource owner mappers map the data received from the OAuth2 authorization server to user information in the Repository.
+
+Resource owner mappers must implement the `Ibexa\Contracts\OAuth2Client\ResourceOwner\ResourceOwnerMapper` interface.
+Four implementations of `ResourceOwnerMapper` are proposed by default:
+
+- `ResourceOwnerToExistingUserMapper` is the base class extended by the following mappers:
+    - `ResourceOwnerIdToUserMapper` doesn't create a new user, but loads a user (resource owner) based on their identifier.
```
Comment (julitafalcondusza) [inline docs/users/oauth_client.md:68]:
```suggestion
    - `ResourceOwnerIdToUserMapper` - loads a user (resource owner) based on the identifier, but doesn't create a new user.
```

File: docs/users/oauth_client.md [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+    - `ResourceOwnerEmailToUserMapper` doesn't create a new user, but loads a user (resource owner) based on their email.
```
Comment (julitafalcondusza) [inline docs/users/oauth_client.md:69]:
```suggestion
    - `ResourceOwnerEmailToUserMapper` - loads a user (resource owner) based on the email, but doesn't create a new user.
```

File: docs/users/oauth_client.md [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+- `ResourceOwnerToExistingOrNewUserMapper` checks if the user exists and loads them if they do.
```
Comment (julitafalcondusza) [inline docs/users/oauth_client.md:70]:
```suggestion
- `ResourceOwnerToExistingOrNewUserMapper` - checks whether the user exists and loads the data if it does.  If not, creates a new user in the Repository.
```

File: docs/users/oauth_client.md [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+  If they don't, the mapper creates a new user in the Repository.
```
Comment (julitafalcondusza) [inline docs/users/oauth_client.md:71]:
```suggestion
```

File: docs/users/oauth_client.md [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+
+To use `ResourceOwnerToExistingOrNewUserMapper`, you need to extend it in your custom mapper.
+
+!!! tip "OAuth User content type"
+
+    When you implement your own mapper for external login,
+    it is good practice to create a special User content type for users registered in this way.
+    This is because users who register through an external service do not have a separate password in the system.
```
Comment (julitafalcondusza) [inline docs/users/oauth_client.md:79]:
```suggestion
    The users who register through an external service do not have a separate password in the system.
```

File: docs/users/oauth_client.md [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+Configure the service by using the `ibexa.oauth2_client.resource_owner_mapper` tag to associate it with the `google` client:
+
+``` yaml
+[[= include_file('code_samples/user_management/oauth_google/config/custom_services.yaml', 0, 6) =]]
+```
+
+## Add login button
+
+After you have activated the OAuth2 client for the `admin` SiteAccess,
+here is how to add a **Log in with Google** to the Back Office login form.
```
Comment (julitafalcondusza) [inline docs/users/oauth_client.md:111]:
```suggestion
you need to add a **Log in with Google** to the Back Office login form.
```

**Review by julitafalcondusza — CHANGES_REQUESTED** [review]
Please check my comments.

**Review by julitafalcondusza — APPROVED** [review]

**Pull request merged**